src/swarm/neo/util/StructPacker.d

```diff
@@ -0,0 +1,360 @@
+    // original.
+    bool packUnpack ( S ) ( S s )
+    {
+        void[] buf;
+        pack(S.init, buf);
+        auto unpacked = unpack!(S)(buf);
+        return deepEquals(s, *unpacked);
+    }
+}
+
+// Tests for `pack` and `unpack`.
+unittest
+{
+    struct Simple
+    {
+        int i = 23;
+        float f = 23.23;
+    }
+    test(packUnpack(Simple.init));
+
+    struct Array
+    {
+        int i = 23;
+        mstring str = "hello".dup;
+    }
+    test(packUnpack(Array.init));
+
+    struct Pointer
+    {
+        void* p;
+    }
+    test(packUnpack(Pointer.init));
+
+    struct StaticArray
+    {
+        ubyte[10] a = [0,1,2,3,4,5,6,7,8,9];
+    }
+    test(packUnpack(StaticArray.init));
+
+    struct StaticArrayNested
+    {
+        struct S
+        {
+            int i;
+        }
+
+        S[10] a = [S(0),S(1),S(2),S(3),S(4),S(5),S(6),S(7),S(8),S(9)];
+    }
+    test(packUnpack(StaticArrayNested.init));
+
+    struct Array2d
+    {
+        int i = 23;
+        mstring[] str = ["hello".dup, "how".dup, "amazing".dup];
+    }
+    test(packUnpack(Array2d.init));
+
+    struct Delegate
+    {
+        void delegate ( ) d;
+    }
+    test(packUnpack(Delegate.init));
+
+    struct Union
+    {
+        union U
+        {
+            void* p;
+            int i;
+        }
+
+        U u;
+    }
+    test(packUnpack(Union.init));
+
+    struct Nested
+    {
+        int i = 23;
+        float f = 23.23;
+
+        struct S
+        {
+            int i = 47;
+        }
+
+        S s;
+    }
+    test(packUnpack(Nested.init));
+
+    struct NestedPointer
+    {
+        struct S
+        {
+            void* p;
+        }
+
+        S s;
+    }
+    test(packUnpack(NestedPointer.init));
+
+    // Replicates the real situation that we have in RequestCore.
+    struct Context
+    {
+        struct UserSpecifiedParams
+        {
+            struct Args
+            {
+                mstring channel = "campaign_metadata".dup;
+                hash_t key = 23;
+            }
+
+            Args a;
+            void delegate ( ) notifier;
+        }
+
+        UserSpecifiedParams usp;
+
+        struct SharedWorking
+        {
+            void* p;
+        }
+
+        SharedWorking shared_working;
+    }
+    test(packUnpack(Context.init));
+}
+
+/*******************************************************************************
+
+    Checks whether the specified type can be packed by the above function.
+
+    Params:
+        S = type to check
+
+    Returns:
+        true if the type can be packed
+
+*******************************************************************************/
+
+private bool canPack ( S ) ( )
+{
+    bool error;
+
+    foreach ( i, F; typeof(S.tupleof) )
+    {
+        // Arrays are allowed, if...
+        static if ( isDynamicArrayType!(F) )
+        {
+            // ...they only contain primitive types.
+            static if ( !isPrimitiveType!(BaseTypeOfArrays!(F)) )
+                error = true;
```

Comment:
- For the sake of an error message that is more comprehensive than "something is wrong" I suggest using `static assert`s on the error conditions with the unsupported type and maybe even the field name in the message, and naming the function `void checkPackable` or something similar.
- Why are dynamic arrays of pointers/delegates/objects etc. disallowed, again, while static arrays of such types are allowed?

Reply:
>Why are dynamic arrays of pointers/delegates/objects etc. disallowed, again, while static arrays of such types are allowed?

Because dynamic array contents are copied to the end of the packed buffer and may thus become misaligned.

Reply:
Ah, right. Now I remember.

Reply:
I'll add a comment explaining this.